kafka/producer.go

```diff
@@ -17,4 +17,198 @@
+type ProducerConfig struct {
+	// Broker holds the (host:port) address of the Kafka broker to which
+	// events should be published.
+	Broker string
+
+	// ClientID to use when connecting to Kafka. This is used for logging
+	// and client identification purposes.
+	ClientID string
+	// Version is the software version to use in the Kafka client. This is
+	// useful since it shows up in Kafka metrics and logs.
+	Version string
+
+	// TopicPrefix holds the name of the Kafka topic prefix to which events
+	// should be published. The producer will send messages to 5 different
+	// topics using the <TopicPrefix>.<EventType> format.
+	TopicPrefix string
+
+	// Logger is used for logging producer errors.
+	Logger *zap.Logger
+
+	// Key, if non-nil, is used to set the key for an event message.
+	//
+	// Producer uses murmur2 partitioning: consistent hashing over the
+	// message key, if it is non-nil; random partition assignment if the
+	// message has no key.
+	//
+	// If Key is nil, the message key will be set to the event's trace ID
+	// if there is one, and nil otherwise.
+	Key func(model.APMEvent) []byte
+
+	// TODO(axw) Kafka credentials
+
+	// Kafka topics to publish to.
+	txTopic     string
+	spanTopic   string
+	metricTopic string
+	errorTopic  string
+	logTopic    string
+}
+
+// Validate checks that cfg is valid, and returns an error otherwise.
+func (cfg ProducerConfig) Validate() error {
+	if cfg.Broker == "" {
+		return errors.New("broker cannot be empty")
+	}
+	if cfg.TopicPrefix == "" {
+		return errors.New("topic cannot be empty")
+	}
+	if cfg.Logger == nil {
+		return errors.New("logger cannot be nil")
+	}
+	return nil
+}
+
+// Producer is a model.BatchProcessor that publishes events to Kafka.
+type Producer struct {
+	cfg    ProducerConfig
+	client *kgo.Client
+
+	mu sync.RWMutex
+}
+
+// NewProducer returns a new Producer with the given config.
+func NewProducer(cfg ProducerConfig) (*Producer, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid producer config: %w", err)
+	}
+
+	opts := []kgo.Opt{
+		kgo.SeedBrokers(cfg.Broker),
+		kgo.WithLogger(kzap.New(cfg.Logger)),
+	}
+	if cfg.ClientID != "" {
+		opts = append(opts, kgo.ClientID(cfg.ClientID))
+		if cfg.Version != "" {
+			opts = append(opts, kgo.SoftwareNameAndVersion(
+				cfg.ClientID, cfg.Version,
+			))
+		}
+	}
+	// TODO(marclop) block on re-balances, auto-commit high watermarks.
+	client, err := kgo.NewClient(opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating producer: %w", err)
+	}
+	// Issue a metadata refresh request on construction, so the broker list is
+	// populated.
+	client.ForceMetadataRefresh()
+
+	cfg.txTopic = fmt.Sprintf("%s.transaction", cfg.TopicPrefix)
+	cfg.spanTopic = fmt.Sprintf("%s.span", cfg.TopicPrefix)
+	cfg.metricTopic = fmt.Sprintf("%s.metric", cfg.TopicPrefix)
+	cfg.errorTopic = fmt.Sprintf("%s.error", cfg.TopicPrefix)
+	cfg.logTopic = fmt.Sprintf("%s.log", cfg.TopicPrefix)
```

Comment:
I'm not sure how I feel about having this here, I think this may not belong in this repository. Instead, I think it would be best to have a list of record mutators that can be passed as configuration. This would provide a cleaner separation between business and implementation logic:

```go
type RecordMutator func(model.APMEvent, *kgo.Record) error

// ProducerConfig holds configuration for publishing events to Kafka.
type ProducerConfig struct {
	// ...
	Mutators []RecordMutator
	// ...
}
```

Not all producers will always produce using the same topics, and we may want to change that independently too in the future (and I believe even now).

Reply:
Good point! Updated :+1: